### PR TITLE
Chore: Apply literal properties 

### PR DESCRIPTION
**Pull request by binarygit**
This PR replaces the initialize method on components to use the `prop` method from the literal gem which provides a way to specify type.

# Description

Fixes https://github.com/avo-hq/avo/issues/2974
Fixes #2893 

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/avo-hq/avodocs)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Manual review steps

1. Step 1
1. Step 2

Manual reviewer: please leave a comment with output from the test if that's the case.

### CODE REVIEW

File: app/components/avo/fields/common/key_value_component.rb [hunk 1]
```diff
@@ -5,9 +5,13 @@ class Avo::Fields::Common::KeyValueComponent < Avo::BaseComponent
...
-    @view = Avo::ViewInquirer.new(view)
+  prop :field, Avo::Fields::BaseField
+  prop :form, _Nilable(ActionView::Helpers::FormBuilder)
+  prop :view, Symbol, default: :show do |value|
+    value&.to_sym
+  end
+
+  def after_initialize
+    @view = Avo::ViewInquirer.new(@view)
   end
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/key_value_component.rb:16]: Can we merge these?

File: app/components/avo/fields/common/progress_bar_component.rb [hunk 2]
```diff
@@ -1,17 +1,15 @@
...
+  end
 
-  def initialize(value:, display_value: false, value_suffix: nil, max: 100, view: "index")
-    @value = value
-    @display_value = display_value
-    @value_suffix = value_suffix
-    @max = max
+  def after_initialize
     @view = Avo::ViewInquirer.new(view)
   end
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/progress_bar_component.rb:14]: Can we merge the `view` initialization and make it as `Avo::ViewInquirer` type?

File: app/components/avo/modal_component.rb [hunk 3]
```diff
@@ -4,13 +4,7 @@ class Avo::ModalComponent < Avo::BaseComponent
   renders_one :heading
   renders_one :controls
 
-  attr_reader :width
-  attr_reader :body_class
-
-  def initialize(width: :md, body_class: nil)
-    @width = width
-    @body_class = body_class
-  end
```
Comment (Paul-Bob) [inline app/components/avo/modal_component.rb:13]: Why are we deleting `@body_class`?

File: app/components/avo/panel_component.rb [hunk 4]
```diff
@@ -18,18 +17,20 @@ class Avo::PanelComponent < Avo::BaseComponent
...
+  prop :data, Hash, default: {}.freeze
+  prop :display_breadcrumbs, _Boolean, default: false
+  prop :index, _Nilable(Integer)
+  prop :classes, _Nilable(String)
+  prop :profile_photo, _Nilable(Avo::ProfilePhoto)
+  prop :cover_photo, _Nilable(Avo::CoverPhoto)
+  prop :args, Hash, :**, default: {}.freeze
+
+  def after_initialize
+    @title = @args[:title]
```
Comment (Paul-Bob) [inline app/components/avo/panel_component.rb:32]: Why not convert the title to a prop?

File: app/helpers/avo/application_helper.rb [hunk 5]
```diff
@@ -12,6 +12,9 @@ def render_license_warning(title: "", message: "", icon: "exclamation")
     end
 
     def empty_state(**args)
+      if args[:view_type]
+        args[:view_type] = args[:view_type].to_sym
+      end
```
Comment (Paul-Bob) [inline app/helpers/avo/application_helper.rb:17]: Why is this necessary?

File: app/views/avo/base/_new_via_belongs_to.html.erb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <%= turbo_frame_tag "new_via_belongs_to" do %>
-  <%= render(Avo::ModalComponent.new(width: :xl, body_class: "bg-application")) do |c| %>
+  <%= render(Avo::ModalComponent.new(width: :xl)) do |c| %>
```
Comment (Paul-Bob) [inline app/views/avo/base/_new_via_belongs_to.html.erb:2]: Why are we removing the `, body_class: "bg-application"`?

**Review by Paul-Bob — CHANGES_REQUESTED** [review]
I left some questions.

I noticed that we have many props with `reader: :public` where the reader is not necessarily needed as public. I know this PR respects the previous `attr_reader` and converts it to `reader: :public` but using the instance variable is faster than accessing through the reader. 

Whenever there is a case where the reader is not necessary since the attribute is used only on the object's scope lets remove the reader and use the instance variable in favor.

File: app/components/avo/modal_component.rb (hunk 3 above)
Comment (binarygit) [reply app/components/avo/modal_component.rb]: This was a mistake. I thought it wasn't being used anywhere until you pointed it out here. I have added this back.

File: app/components/avo/panel_component.rb (hunk 4 above)
Comment (binarygit) [reply app/components/avo/panel_component.rb]: I have converted this and other places where we use `after_initialize` to props

File: app/helpers/avo/application_helper.rb (hunk 5 above)
Comment (binarygit) [reply app/helpers/avo/application_helper.rb]: I have reverted this

File: app/components/avo/fields/common/progress_bar_component.rb (hunk 2 above)
Comment (binarygit) [reply app/components/avo/fields/common/progress_bar_component.rb]: done. Thanks for the suggestion

File: app/components/avo/button_component.rb [hunk 7]
```diff
@@ -4,24 +4,21 @@ class Avo::ButtonComponent < Avo::BaseComponent
...
+    value&.to_sym
+  end
+  prop :icon_class, String, default: ""
+  prop :is_link, _Boolean, default: false
+  prop :rounded, _Boolean, default: true
+  prop :compact, _Boolean, default: false
+  prop :aria, Hash, default: {}.freeze
+  prop :args, Hash, :**, default: {}.freeze
+  prop :class, _Nilable(String) do |value|
+    @args[:class] if @args
```
Comment (Paul-Bob) [inline app/components/avo/button_component.rb:21]: Is `@args[:class] if @args` necessary?

File: app/components/avo/fields/common/badge_viewer_component.rb [hunk 8]
```diff
@@ -1,9 +1,9 @@
 # frozen_string_literal: true
 
 class Avo::Fields::Common::BadgeViewerComponent < Avo::BaseComponent
-  def initialize(value:, options:)
-    @value = value
-    @options = options
+  prop :value, String
+  prop :options, Hash
+  prop :backgrounds, _Nilable(Hash) do |value|
     @backgrounds = {
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/badge_viewer_component.rb:7]: Let's put `@backgrounds` initialization inside `after_initialize`

```suggestion
  def after_initialize
    @backgrounds = {
```

File: app/components/avo/fields/common/files/view_type/list_item_component.rb [hunk 9]
```diff
@@ -1,6 +1,11 @@
 # frozen_string_literal: true
 
 class Avo::Fields::Common::Files::ViewType::ListItemComponent < Avo::Fields::Common::Files::ViewType::GridItemComponent
+  prop :field, Avo::Fields::BaseField
+  prop :resource, Avo::BaseResource
+  prop :file, _Nilable(ActiveStorage::Attachment)
+  prop :extra_classes, _Nilable(String)
+
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/files/view_type/list_item_component.rb:8]: Aren't those already inherited from `Avo::Fields::Common::Files::ViewType::GridItemComponent`?

```suggestion
```

File: app/components/avo/fields/common/key_value_component.rb [hunk 10]
```diff
@@ -3,11 +3,10 @@
...
-  attr_reader :view
-
-  def initialize(field:, form: nil, view: "show")
-    @field = field
-    @form = form
-    @view = Avo::ViewInquirer.new(view)
+  prop :field, Avo::Fields::BaseField
+  prop :form, _Nilable(ActionView::Helpers::FormBuilder)
+  prop :view, Avo::ViewInquirer do |value|
+    value = :show if value.nil?
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/key_value_component.rb:9]: Can we use `default` here?

```suggestion
  prop :view, Avo::ViewInquirer, default: :show do |value|
    value = :show if value.nil?
```

File: app/components/avo/fields/common/progress_bar_component.rb [hunk 11]
```diff
@@ -1,18 +1,13 @@
...
-    @display_value = display_value
-    @value_suffix = value_suffix
-    @max = max
-    @view = Avo::ViewInquirer.new(view)
+  prop :value, Integer
+  prop :display_value, _Boolean, default: false
+  prop :value_suffix, _Nilable(String)
+  prop :max, Integer, default: 100
+  prop :view, Avo::ViewInquirer, reader: :public do |value|
+    value = :index if value.nil?
```
Comment (Paul-Bob) [inline app/components/avo/fields/common/progress_bar_component.rb:9]: Can we use `default` here?

```suggestion
  prop :view, Avo::ViewInquirer, reader: :public, default: :index do |value|
    value = :index if value.nil?
```

File: app/components/avo/index/field_wrapper_component.rb [hunk 12]
```diff
@@ -1,17 +1,19 @@
...
+  prop :center_content, _Boolean, default: false
+  prop :flush, _Boolean, default: false
+  prop :args, Hash, :**, default: {}.freeze
+  prop :view, _Nilable(Avo::ViewInquirer), default: Avo::ViewInquirer.new("index").freeze
+  prop :classes, _Nilable(String) do |value|
+    if @args && @args[:class].present?
+      @args[:class]
+    else
+      ""
+    end
```
Comment (Paul-Bob) [inline app/components/avo/index/field_wrapper_component.rb:16]:
```suggestion
    @args&.dig(:class) || ""
```

File: app/components/avo/panel_component.rb [hunk 13]
```diff
@@ -18,18 +15,21 @@ class Avo::PanelComponent < Avo::BaseComponent
...
+  prop :classes, _Nilable(String)
+  prop :profile_photo, _Nilable(Avo::ProfilePhoto)
+  prop :cover_photo, _Nilable(Avo::CoverPhoto)
+  prop :args, Hash, :**, default: {}.freeze
+  prop :name, _Nilable(String) do |value|
+    if @args
+      @args[:title]
+    else
+      value
+    end
```
Comment (Paul-Bob) [inline app/components/avo/panel_component.rb:32]:
```suggestion
    @args&.dig(:title) || value
```

File: app/components/avo/views/resource_edit_component.rb [hunk 14]
```diff
@@ -3,13 +3,16 @@
 class Avo::Views::ResourceEditComponent < Avo::ResourceComponent
   include Avo::ApplicationHelper
 
-  attr_reader :actions, :display_breadcrumbs
+  prop :resource, _Nilable(Avo::BaseResource)
+  prop :record, _Nilable(ActiveRecord::Base)
+  prop :actions, _Array(Avo::BaseAction), default: [].freeze
+  prop :view, Avo::ViewInquirer do |value|
+    value = :edit if value.nil?
```
Comment (Paul-Bob) [inline app/components/avo/views/resource_edit_component.rb:10]: Can we use `default` here?
```suggestion
  prop :view, Avo::ViewInquirer, default: :edit do |value|
```

File: app/views/avo/base/_new_via_belongs_to.html.erb (hunk 6 above)
Comment (Paul-Bob) [inline app/views/avo/base/_new_via_belongs_to.html.erb:2]: Let's revert this change
```suggestion
  <%= render(Avo::ModalComponent.new(width: :xl, body_class: "bg-application")) do |c| %>
```

**Review by Paul-Bob — COMMENTED** [review]
I left some questions and some tweaks comments, overall is looking great, thank you for changing the readers to ivars!

File: app/components/avo/button_component.rb (hunk 7 above)
Comment (binarygit) [reply app/components/avo/button_component.rb:21]: without the `if @args` we'll get an NoMethodError for `[:class] method doesn't exist for nil`

File: app/components/avo/button_component.rb [hunk 15]
```diff
@@ -4,24 +4,21 @@ class Avo::ButtonComponent < Avo::BaseComponent
...
   end
```
Comment (Paul-Bob) [inline app/components/avo/button_component.rb:22]:
```suggestion
  prop :class, _Nilable(String)
```

**Review by Paul-Bob — APPROVED** [review]
Thank you for the effort on this one @binarygit!

**Comment by xeron** [comment]
@binarygit @Paul-Bob this broke my custom code for setting current user based on HTTP basic auth:

`/app/avo/concerns/avo_current_user.rb`:

```ruby
module AvoCurrentUser
  extend ActiveSupport::Concern

  User = Struct.new(:name)

  included do
    before_action :set_paper_trail_whodunnit
  end

  protected

    def user_for_paper_trail
      current_user.name
    end

    def current_user
      user, _pass = ActionController::HttpAuthentication::Basic.user_name_and_password(request)
      User.new(user)
    end
end
```

`/config/initializers/avo.rb`:

```ruby
Rails.configuration.to_prepare do
  Avo::ApplicationController.include AvoCurrentUser
end
```

File: app/components/avo/fields/common/files/controls_component.rb [hunk 16]
```diff
@@ -4,16 +4,13 @@ class Avo::Fields::Common::Files::ControlsComponent < Avo::BaseComponent
...
-  delegate :id, to: :field
+  delegate :id, to: :@field
 
-  def initialize(field:, file:, resource:)
-    @field = field
-    @file = file
-    @resource = resource
-  end
+  prop :field, Avo::Fields::BaseField
+  prop :file, ActiveStorage::Attachment
```
Comment (segiddins) [inline app/components/avo/fields/common/files/controls_component.rb:10]: This breaks eager-loading on rails apps that dont have `active_storage/engine` required

Comment (adrianthedev) [reply app/components/avo/fields/common/files/controls_component.rb:10]: Thanks! We missed it before.
Fixed by https://github.com/avo-hq/avo/pull/3215/
